packages/react-components/react-teaching-popover-preview/src/components/TeachingPopoverButton/useTeachingPopoverButton.ts

```diff
@@ -1,34 +1,116 @@
+ * @param props - User provided props to the TeachingPopoverButton component.
+ * @param ref - User provided ref to be passed to the TeachingPopoverButton component.
  */
 export const useTeachingPopoverButton_unstable = (
   props: TeachingPopoverButtonProps,
-  ref: React.Ref<HTMLDivElement>,
+  ref: React.Ref<HTMLButtonElement | HTMLAnchorElement>,
 ): TeachingPopoverButtonState => {
-  return {
-    // TODO add appropriate props/defaults
-    components: {
-      // TODO add each slot's element type or component
-      root: 'div',
+  const { buttonType, altStepText } = props;
+  const toggleOpen = usePopoverContext_unstable(context => context.toggleOpen);
+  const triggerRef = usePopoverContext_unstable(context => context.triggerRef);
+
+  const totalPages = useTeachingPopoverContext_unstable(context => context.totalPages);
+  const currentPage = useTeachingPopoverContext_unstable(context => context.currentPage);
+  const setCurrentPage = useTeachingPopoverContext_unstable(context => context.setCurrentPage);
+  const appearance = useTeachingPopoverContext_unstable(context => context.appearance);
+  const onFinish = useTeachingPopoverContext_unstable(context => context.onFinish);
+  const onPageChange = useTeachingPopoverContext_unstable(context => context.onPageChange);
+
+  const isCarousel = totalPages > 1;
+  let buttonAppearanceMod = {};
+  let buttonTextMod = {};
+
+  if (!props.appearance) {
+    if (buttonType === 'primary') {
+      const shouldShowFinalStepText = altStepText && currentPage === totalPages - 1 && totalPages > 1;
+      buttonTextMod = shouldShowFinalStepText ? { children: altStepText } : {};
+      buttonAppearanceMod = isCarousel ? { appearance: 'primary' } : {};
+      if (appearance === 'brand') {
+        buttonAppearanceMod = isCarousel ? {} : { appearance: 'primary' };
+      }
+    } else {
+      const shouldShowInitialStepText = altStepText && currentPage === 0 && totalPages > 1;
+      buttonTextMod = shouldShowInitialStepText ? { children: altStepText } : {};
+      buttonAppearanceMod = isCarousel ? {} : { appearance: 'primary' };
+      if (appearance === 'brand') {
+        buttonAppearanceMod = isCarousel ? { appearance: 'outline' } : {};
+      }
+    }
+  }
+
+  if (altStepText && isCarousel) {
+    /* Alternate text for first/last step depending on button type */
+    if (buttonType === 'primary') {
+      const shouldShowFinalStepText = altStepText && currentPage === totalPages - 1;
+      buttonTextMod = shouldShowFinalStepText ? { children: altStepText } : {};
+    } else {
+      const shouldShowInitialStepText = altStepText && currentPage === 0;
+      buttonTextMod = shouldShowInitialStepText ? { children: altStepText } : {};
+    }
+  }
+
+  const handleTeachingPopoverButtonClick = useEventCallback(
+    (event: React.MouseEvent<HTMLButtonElement & HTMLAnchorElement & HTMLDivElement>) => {
+      props.onClick?.(event);
+
+      if (!event.isDefaultPrevented()) {
```

Comment:
More a style comment but you can eliminate one level of indentation by change this is 

```ts
if (event.isDefaultPrevented()) {
  return;
}
```